torch/utils/data/dataset.py

```diff
@@ -316,6 +316,8 @@ def __init__(self, dataset: Dataset[T_co], indices: Sequence[int]) -> None:
         self.indices = indices
 
     def __getitem__(self, idx):
+        if isinstance(self.indices, list) and isinstance(idx, Iterable):
```

Comment:
Let's change it to `if isinstance(idx, list):`. Let me know if you have any objection.

Reply:
Simply removing the `isinstance(idx, Iterable)` will cause error, this is because `self.indices` may be a list but `idx` will not necessarily be an iterable. Passing a non iterable `idx` to the for loop in line 320 will cause error. E.g. slicing a subset  `subset[0:-1]` will give a sice object, `slice(0, -1, None)`, as `idx`  , which is not iterable. Indexing a subset would also give error since `subset[0]` will set ` idx` as 0 and this will also cause error in the line 320 for loop. But these non iterable `idx` can be handled by the original `return self.dataset[self.indices[idx]]`.

The feature line 319 and 320 add is the ability to index a subset by an Iterable E.g. `subset[0,1,2,3,4] `. This would have return an error in the previous implementation, and was causing the error when trying to slice a subset of a subset. E.g. if `idx` is a slice, `self.indices[idx]` is a tuple, an Iterable, hence line 321 would be `return self.dataset[**Iterable**]`. This would be ok if `self.dataset` is a Dataset since it can be indexed by an Iterable, but if `self.dataset` is a Subset then it will return error. Hence adding this feature solves the original problem of "Subset of a Subset not sliceable issue".

Sorry for the long comment, I hope it helps. Please let me know if there are any problems or changes that should be made. Thank you.

Reply:
I am referring the condition that `idx` is a `list` (not `indices` is `list`). We should expect `indices` always being `list`.

Reply:
In general, I mean we can delete the check for `isinstance(self.indices, list)`, which should always be true.

Reply:
I see, I am sorry for the miss understanding. The reason why I left the list there is because I thought someone may want to use a numpy array or integer pytorch tensor as an indices instead, which already supports being indexed by an Iterable, and they may be faster than doing a python for loop. But sure, I don't have any other reasonable objection against removing it so I am very much happy to remove it.

Reply:
IMO, the reason is kind simple. Subset is normally used by `random_split`. If users want to use different ways to create Subset, it can be easily achieved to write a custom class. It's not possible for us to support all the user behavior for a single class.
BTW, please also switch to check `idx` against `list` rather than `Iterable`. If `idx` is a tuple, it should not fall into this logic.

Reply:
Sure, I completely agree with you. I had a wrong idea on how indexing by tuple behaves on a torch tensors. I think the interface of Subset should be similar to of Dataset with a polymorphism like concept.